specs/phase1/beacon-chain.md

```diff
@@ -1057,51 +532,124 @@ def process_epoch(state: BeaconState) -> None:
     process_reveal_deadlines(state)
     process_challenge_deadlines(state)
     process_slashings(state)
+    process_pending_headers(state)
+    charge_confirmed_header_fees(state)
+    reset_pending_headers(state)
```

Comment:
Is there a particular reason to process pending headers during epoch transition?

Is it assumed that there will be a relatively low number of pending headers?

Probably a premature optimization, but maybe it makes sense to process pending headers more often during the epoch instead of already quite busy epoch processing.

Reply:
Technically, they already _are_ being processed, in that if a header gets to 2/3 during an epoch it gets confirmed immediately. The end-of-epoch processing is just the case where no header gets to 2/3. I suppose if we want, we could put it into per-slot processing with a delay of 32 slots, but that may complicate the code more....

Reply:
I would say that the main question is do we want to offload pending headers processing from epoch processing. If yes, then not so complicated code change would be to move pending headers processing from the epoch processing to let's say a slot close to the beginning of an epoch. Then there would not be a need to complicate the code with a moving window approach. 

My main argument now would be that here we are dealing with shards (not the main beacon chain) that could become pretty chaotic during turbulent periods. But I agree that it's too early to complicate the code here until we have some real numbers.

Reply:
Just to clarify, by "pending headers" I mean the headers that are still not processed (not received 2/3 votes) at the end-of-epoch. It will be interesting to see how much of those actually piles up and what actual overhead it brings to the epoch processing.

Reply:
Hmm. I would be inclined not to worry too much, especially since HF1 is reducing end-of-epoch work quite a bit anyway, and the amount of work being done here is significantly less than the rest of attestation processing.

Reply:
I guess we can just wait until we see the real numbers. Testnets (usually with way lower participation rates comparing to the mainnet) should show is there any real need to move pending headers processing from the end-of-epoch.

But I would like to see a more general rule that processing is spread out uniformly across entire epoch, especially for low hanging fruits such as offloading processing from the end-of-epoch if it doesn't need to be processed at the end-of-epoch. This would help client teams to deliver faster as this approach should be less demanding for optimizations. This would also help to avoid hard forks or will make it simpler because seems that accumulating processing at specific points (like end-of-epoch) sooner or later needs a hard fork.